### PR TITLE
mgr/dashboard: Manager should complain about wrong dashboard certificate

### PR DESCRIPTION
**Pull request by votdev**
Fixes: https://tracker.ceph.com/issues/24453

Signed-off-by: Volker Theile <vtheile@suse.com>

- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
IMO we should not do this. It's getting security sensitive, especially if it comes to not validating stuff like intermediate certificates. I wouldn't add this just for printing some warnings, unless it is needed to provide proper security.

**Comment by LenzGr** [comment]
> IMO we should not do this. It's getting security sensitive, especially if it comes to not validating stuff like intermediate certificates. I wouldn't add this just for printing some warnings, unless it is needed to provide proper security.

I think this an improvement over the current situation, in which we silently fail if the certificate does not match the expected format, leaving the user puzzled over what's going on.

**Comment by votdev** [comment]
Maybe we should only take care that the files are in a correct format and skip all other checks like checking the expiration date and private key and certificate do match up.

**Comment by sebastian-philipp** [comment]
> > IMO we should not do this. It's getting security sensitive, especially if it comes to not validating stuff like intermediate certificates. I wouldn't add this just for printing some warnings, unless it is needed to provide proper security.
> 
> I think this an improvement over the current situation, in which we silently fail if the certificate does not match the expected format, leaving the user puzzled over what's going on.

What do you mean by "silently fail", do we print a user visible message in the log or module status?

**Comment by LenzGr** [comment]
> What do you mean by "silently fail", do we print a user visible message in the log or module status?

See the corresponding [bug report](https://tracker.ceph.com/issues/24453):

> Configuring the dashboard and selecting certificates didn't work as intended.
> Dashboard wouldn't load (it didn't even listen on any ports).
> When reverted to using self-signed certificates, dashboard suddenly works.
> Expected behaviour: if, for whatever reason, certificates are wrong (format, type, ...) mgr should throw an error in log.
> Actual behaviour: manager was silently failing to display dashboard

**Comment by LenzGr** [comment]
> Maybe we should only take care that the files are in a correct format and skip all other checks like checking the expiration date and private key and certificate do match up.

If possible, I'd suggest to abort with an error if the format is incorrect, and log warnings about other inconsistencies like expired certificates and key mismatches.

**Comment by sebastian-philipp** [comment]
A better alternative would be to fix the logging by properly setting Python the default logger instead. Maybe with an approach like this: https://stackoverflow.com/questions/7428426/redefining-logging-root-logger

**Comment by votdev** [comment]
> A better alternative would be to fix the logging by properly setting Python the default logger instead. Maybe with an approach like this: [stackoverflow.com/questions/7428426/redefining-logging-root-logger](https://stackoverflow.com/questions/7428426/redefining-logging-root-logger)

I configured the root logger the following way, but did not get any log message when i try to use an invalid SSL certificate.
```
import logging
root_logger = logging.getLogger()
root_logger.setLevel(logging.DEBUG)
root_logger.addHandler(CPlusPlusHandler(self))
```
So it seems CherryPy nor the used SSL backend does log invalid keys/certificates.

**Comment by sebastian-philipp** [comment]
with the latest cherrypy and enabled logging, I'm getting:

```
ERROR:cherrypy.error:[19/Mar/2019:14:14:35] ENGINE Error in 'start' listener <bound method Server.start of <cherrypy._cpserver.Server object at 0x7fe082a49e10>>
Traceback (most recent call last):
  File "/tmp/foo/venv/lib/python3.6/site-packages/cherrypy/process/wspbus.py", line 230, in publish
    output.append(listener(*args, **kwargs))
  File "/tmp/foo/venv/lib/python3.6/site-packages/cherrypy/_cpserver.py", line 179, in start
    self.httpserver, self.bind_addr = self.httpserver_from_self()
  File "/tmp/foo/venv/lib/python3.6/site-packages/cherrypy/_cpserver.py", line 170, in httpserver_from_self
    httpserver = _cpwsgi_server.CPWSGIServer(self)
  File "/tmp/foo/venv/lib/python3.6/site-packages/cherrypy/_cpwsgi_server.py", line 103, in __init__
    self.server_adapter.ssl_ciphers)
  File "/tmp/foo/venv/lib/python3.6/site-packages/cheroot/ssl/builtin.py", line 104, in __init__
    self.context.load_cert_chain(certificate, private_key)
ssl.SSLError: [X509: KEY_VALUES_MISMATCH] key values mismatch (_ssl.c:3507)

[19/Mar/2019:14:14:35] ENGINE Shutting down due to error in start listener:
Traceback (most recent call last):
  File "/tmp/foo/venv/lib/python3.6/site-packages/cherrypy/process/wspbus.py", line 268, in start
    self.publish('start')
  File "/tmp/foo/venv/lib/python3.6/site-packages/cherrypy/process/wspbus.py", line 248, in publish
    raise exc
cherrypy.process.wspbus.ChannelFailures: SSLError(185073780, '[X509: KEY_VALUES_MISMATCH] key values mismatch (_ssl.c:3507)')

```

still investigating ....

**Comment by votdev** [comment]
After doing some research it seems that the SSL library does not log errors, it only throws exceptions (see @sebastian-philipp comment above). We could catch this exceptions and log the exception message, but this is a meaningless message like ``[X509: KEY_VALUES_MISMATCH] key values mismatch (_ssl.c:3507)``. Because of that i vote for doing the checks ourself and log meaningful messages.

**Comment by sebastian-philipp** [comment]
I'd be +1 for enabling the  logging within `HTTPServer.tick` in any case

**Comment by votdev** [comment]
> I'd be +1 for enabling the logging within `HTTPServer.tick` in any case

See #27069

File: src/pybind/mgr/dashboard/module.py [hunk 1]
```diff
@@ -197,6 +199,33 @@ def _configure(self):
...
+            # Do some validations to the private key and certificate:
+            # - Check the type and format
+            # - Check the certificate expiration date
+            # - Check the consistency of the private key
+            # - Check that the private key and certificate match up
+            try:
+                with open(cert_fname) as f:
+                    x509 = crypto.load_certificate(crypto.FILETYPE_PEM, f.read())
+                    if x509.has_expired():
+                        self.log.warning('Certificate has been expired')
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/module.py]: does it make sense to report the path of cert in the exception and/or in the log message? like, print out the `cert_fname` ?

the same applies to `pkey_fname`.

Comment (tchaikov) [reply src/pybind/mgr/dashboard/module.py]: also, to improve the readability, how about extracting this part into a static method?

Comment (votdev) [reply src/pybind/mgr/dashboard/module.py]: Done

**Pull request merged**